packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -354,36 +354,40 @@ class AndroidDevice extends Device {
       // device has printed "Observatory is listening on...".
       printTrace('Waiting for observatory port to be available...');
 
+      // TODO(danrubel): The iOS device class does something similar to this code below.
+      // The various Device subclasses should be refactored and common code moved into the superclass.
       try {
-        int observatoryDevicePort, diagnosticDevicePort;
+        Uri observatoryDeviceUri, diagnosticDeviceUri;
 
         if (debuggingOptions.buildMode == BuildMode.debug) {
-          Future<List<int>> scrapeServicePorts = Future.wait(
-              <Future<int>>[observatoryDiscovery.nextPort(), diagnosticDiscovery.nextPort()]
+          Future<List<Uri>> scrapeServiceUris = Future.wait(
+              <Future<Uri>>[observatoryDiscovery.nextUri(), diagnosticDiscovery.nextUri()]
           );
-          List<int> devicePorts = await scrapeServicePorts.timeout(new Duration(seconds: 20));
-          observatoryDevicePort = devicePorts[0];
-          diagnosticDevicePort = devicePorts[1];
+          List<Uri> deviceUris = await scrapeServiceUris.timeout(new Duration(seconds: 20));
+          observatoryDeviceUri = deviceUris[0];
+          diagnosticDeviceUri = deviceUris[1];
         } else {
-          observatoryDevicePort = await observatoryDiscovery.nextPort().timeout(new Duration(seconds: 20));
+          observatoryDeviceUri = await observatoryDiscovery.nextUri().timeout(new Duration(seconds: 20));
         }
 
-        printTrace('observatory port on device: $observatoryDevicePort');
+        printTrace('Observatory Uri on device: $observatoryDeviceUri');
         int observatoryLocalPort = await debuggingOptions.findBestObservatoryPort();
         // TODO(devoncarew): Remember the forwarding information (so we can later remove the
         // port forwarding).
-        observatoryLocalPort = await _forwardPort(ProtocolDiscovery.kObservatoryService, observatoryDevicePort, observatoryLocalPort);
-
-        int diagnosticLocalPort;
-        if (diagnosticDevicePort != null) {
-          printTrace('diagnostic port on device: $diagnosticDevicePort');
-          diagnosticLocalPort = await debuggingOptions.findBestDiagnosticPort();
-          diagnosticLocalPort = await _forwardPort(ProtocolDiscovery.kDiagnosticService, diagnosticDevicePort, diagnosticLocalPort);
+        observatoryLocalPort = await _forwardPort(observatoryLocalPort, observatoryDeviceUri.port);
+        Uri observatoryLocalUri = observatoryDeviceUri.replace(port: observatoryLocalPort);
+
+        Uri diagnosticLocalUri;
+        if (diagnosticDeviceUri != null) {
```

Comment:
so this block is protected by an if statement but the similar block above isn't. I think you can get rid of the if statement here or is it possible to get an Observatory URI but not a Diagnostic URI?